lib/services/local/users.go

```diff
@@ -1782,6 +1782,57 @@ func (s *IdentityService) GetSSODiagnosticInfo(ctx context.Context, authKind str
 	return &req, nil
 }
 
+func (s *IdentityService) UpsertSSOMFASessionData(ctx context.Context, sd *services.SSOMFASessionData) error {
+	switch {
+	case sd == nil:
+		return trace.BadParameter("missing parameter sd")
+	case sd.RequestID == "":
+		return trace.BadParameter("missing parameter RequestID")
+	case sd.ConnectorID == "":
+		return trace.BadParameter("missing parameter ConnectorID")
+	case sd.ConnectorType == "":
+		return trace.BadParameter("missing parameter ConnectorType")
+	case sd.Username == "":
+		return trace.BadParameter("missing parameter Username")
+	}
+
+	value, err := json.Marshal(sd)
```

Comment:
If this is a brand new type I'd strongly recommend protojson, FWIW.

Reply:
The struct contains a protobuf-defined type; is that currently (wrongly) marshaled and unmarshaled with `encoding/json`?

Reply:
I suppose so, what issues does this cause? We are already json marshalling `*mfav1.ChallengeExtensions` in the webauthn session data as well.

Reply:
Do you think a custom marshaller on `services.SSOMFASessionData` to `protojson.Marshal` the challenge extensions is warranted?

Reply:
The `encoding/json` encoding of a protobuf message is not canonical in any way, which means that Go ends up being the only thing that knows what the shape of the json data is supposed to be. Perhaps you could define `SSOMFASessionData` as a protobuf message even if it's not directly used through grpc? That way you also get breaking change detection as part of the linting, and in a future where we generate protobuf types for javascript you'll get the same type definition on that side as well, without having to manually sync it.

Reply:
This session data never leaves the Auth server, so whether or not it's supported across a transport layer doesn't really matter.

Furthermore, I don't see exactly why it would be difficult to maintain the same JSON structure across different languages, but if this is the case and we changed our Auth/Proxy servers to non-go, we'd have a lot of issues with proto fields in our web api json requests and responses. e.g:

```
type getLoginAlertsResponse struct {
	Alerts []types.ClusterAlert `json:"alerts"`
}
```

I don't think we have any plan to have a different language implementation of Teleport, so I don't think this is a problem we need to concern ourselves with anyways, right?

Also, this is just a short lived resource, so we could easily change it in the future without any backwards compatibility concerns. I'll merge this as is as we can always make a follow up PR to change it.

Reply:
It's a resource that gets marshalled and unmarshalled, and it gets stored in a backend. What even _is_ the forward compatible breaking change path for something like that? How do you change it without it resulting in massive breakages?

The fact that we are relying on the implicit data shape of protobuf messages to use `encoding/json` is already very much a big problem, one that for example forces us to keep using the gogoproto generator and library even though we should've stopped using it a long time ago, and each new thing that we marshal like this makes it harder and harder to keep track of what the actual schema of our data is.

> I don't think we have any plan to have a different language implementation of Teleport, so I don't think this is a problem we need to concern ourselves with anyways, right?

We literally face this problem when we aspire to use protobuf in our frontend and are faced with the reality that the user-facing data shape for various things (which matches the data shape stored in our backend) is something vaguely deterministic that's only defined by how a specific code generator for Go decided to arrange things rather than anything canonical - *while also dealing with all the drawbacks of a generated data type*.

Reply:
If that's the only possible failure scenario then sure, it's not a problem - what happens if the data is misinterpreted because a new protobuf codegen has changed something in how the struct is laid out and we didn't notice, tho?

Reply:
Then the value should be treated as the zero value, so it would be treated like a challenge with an unspecified scope with reuse not allowed, which is the least privileged type of challenge and would get rejected in basically any context. Anyways, I'll make a follow up PR.